src/ophyd_async/core/_providers.py

```diff
@@ -17,43 +20,155 @@ class DirectoryInfo:
 
     :param root: Path of a root directory, relevant only for the file writer
     :param resource_dir: Directory into which files should be written, relative to root
-    :param prefix: Optional filename prefix to add to all files
-    :param suffix: Optional filename suffix to add to all files
+    :param filename: Base filename to use generated by FilenameProvider
+    :param create_dir_depth: Optional depth of dirs to create if they do not exist
     """
 
     root: Path
     resource_dir: Path
-    prefix: Optional[str] = ""
-    suffix: Optional[str] = ""
+    filename: str
+    create_dir_depth: Optional[int] = -3
 
 
-class DirectoryProvider(Protocol):
+class FilenameProvider(Protocol):
     @abstractmethod
-    def __call__(self) -> DirectoryInfo:
+    def __call__(self) -> str:
+        """Get a filename to use for output data"""
+
+
+class PathProvider(Protocol):
+    _filename_provider: FilenameProvider
+
+    @abstractmethod
+    def __call__(self, device_name=None) -> PathInfo:
         """Get the current directory to write files into"""
 
 
-class StaticDirectoryProvider(DirectoryProvider):
+class StaticFilenameProvider(FilenameProvider):
+    def __init__(self, filename: str):
+        self._static_filename = filename
+
+    def __call__(self) -> str:
+        return self._static_filename
+
+
+class UUIDFilenameProvider(FilenameProvider):
+    def __init__(self, uuid_call_func: callable = uuid.uuid4):
```

Comment:
May be able to make use of a ParamSpec here to have Callable[P], *calling_args: Optional[P] = None

Not sure how complex it may be however

Then __call__ = callable(calling_args) if calling_args else callable()

Reply:
@DiamondJoseph, what is ParamSpec?

Reply:
Found that: https://docs.python.org/3/library/typing.html#typing.ParamSpec.